### ISSUE TITLE
Swagger + Attachments documentation?

### ISSUE THREAD

**Issue by Prudwell** [couchbaselabs/couchbase-mobile-portal#536]
I'm a bit confused as to how the swagger spec wants me to structure a request to the [put-db-doc-attachment](https://developer.couchbase.com/documentation/mobile/1.3/references/sync-gateway/admin-rest-api/index.html#!/attachment/put_db_doc_attachment) endpoint. 

I'm doing some post-processing of data on the admin port but I don't see how to set the body content. I also don't see the rev parameter in the swagger spec so I'm not sure where to include that. Here's the code I'm working on.

```
let attachObj = {
    db: 'consumer_db_one',
    doc: "generatedFile::" + userID,
    attachment: 'todosExport.txt',
    rev: res.obj.rev 
}
adminClient.attachment.put_db_doc_attachment(attachObj, {requestContentType:"text/plain"})
```

I get 409 conflict error for this code so I don't think I'm setting the rev correctly.

The description in the swagger spec also mentions setting the content type header and it's not clear to me where that's supposed to happen. The above code throws the error "server doesn't consume text/plain, try ["application/json"]" so I'm definitely not doing it right.

I've checked out [this guide](https://developer.couchbase.com/documentation/mobile/1.3/guides/sync-gateway/integrating-external-stores/index.html#importing-attachments) on integrating external stores and it does use swagger for attachment upload but that's on bulk_docs and I'm still not sure how the attachment endpoint itself works. 

I would love love love code samples being put into the swagger specs or some other place someday.

Any ideas how I can learn more about how to work with that endpoint? Thanks!

**Comment by Prudwell** [comment]
Well I've got it working using fetch but I'd really rather not jump back and forth between using swagger client and handwritten fetch requests.

```
let docId = encodeURIComponent("generatedFile::" + userID)
let remote = "http://" + sync_gateway_url + ":4985/app_db/"+docId+"/todosExport.txt?rev=" + encodeURIComponent(res.obj.rev)
let settings ={
    headers: {
        "Content-Type": "text/plain"
    },
    method: 'PUT',
    body: generatedTodoTextDocument
}
// console.log(settings)

fetch(remote, settings)
    .then((res) => {
        // console.log(res)
        return res.text()
    })
    .then((res) => {
        console.log(res)
    })
```

Is the swagger spec just missing parameters for the rev and body in [put_db_doc_attachment](https://developer.couchbase.com/documentation/mobile/1.3/references/sync-gateway/rest-api/index.html#!/attachment/put_db_doc_attachment) or am I misunderstanding how to use it? I see that the [get_db_doc_attachment](https://developer.couchbase.com/documentation/mobile/1.3/references/sync-gateway/admin-rest-api/index.html#!/attachment/get_db_doc_attachment) endpoint does include rev. 

Cheers

**Comment by Prudwell** [comment]
Looks like this got addressed in [594](594), closing.

**Comment by jamiltz** [comment]
@Prudwell as you know, for the most part, the swagger spec is self documented. But we're working on some code samples for the non-trivial tasks. Post-processing on the admin port sounds like a useful scenario to document. Can you share the code snippet that ended up working for you?

**Comment by Prudwell** [comment]
I actually never did get it working with the attachment endpoint; I ended up using the _bulk_docs endpoint in a manner similar to the method described [here](https://developer.couchbase.com/documentation/mobile/1.3/guides/sync-gateway/integrating-external-stores/index.html#importing-attachments). bulk_docs made sense for my use case as often this post processing would generate multiple files each time it ran. I closed my initial Issue with the attachment endpoint when I saw that there was a merge relating to the missing body field, I haven't actually tested that yet.

```
let bulkDocsBody = {
    docs:[],
    new_edits:true
}

let generatedFileObjectCreator = function (idPrefix, filename, body, userID){
    return({
        _id : idPrefix + userID,
        _attachments:
        {
            [filename]: {
                content_type: "text/plain",
                data: btoa(body)
            }
        },
        owner: userID,
        created_at: new Date().getTime(),
        type: "generatedFile"

        })
    }


//...farther along in the code, this happened for several filetypes

let providerObject =  generatedFileObjectCreator('generatedProvidersFile::', "providersExport.txt", generatedDocument, userID)
if(generatedFilesMetaFromDB['generatedProvidersFile::'+ userID]){
    providerObject._rev = generatedFilesMetaFromDB['generatedProvidersFile::'+ userID].rev
}
bulkDocsBody.docs.push(providerObject)


//and finally

adminClient.database.post_db_bulk_docs({db: SYNC_GATEWAY_DATABASE, BulkDocsBody : bulkDocsBody})


```

**Comment by jamiltz** [comment]
Understood, thanks for the details. It looks like swagger js supports media uploads https://github.com/swagger-api/swagger-js/blob/master/test/browser/petstore.js#L17. I'm going to give a go but must fix the spec to have the correct `consumes` parameter.

- [ ] fix swagger spec to be similar to /pet/{petId}/uploadImage
- [ ] try uploading an attachment with the `Blob` api

**Comment by jamiltz** [comment]
Having spent more time on this. I managed to upload an attachment to `/{db}/{id}/{attach}` using the file picker ([image](https://cloud.githubusercontent.com/assets/2589337/23766265/fe150d32-04fb-11e7-9aa8-2dcc3440391d.png)). After inspecting the request/response, the request appears to be the attachment as a binary string. I've been trying to do the same using the JS FileAPI (http://codepen.io/Jamiltz/pen/evgxRm?editors=1011) but no luck so far.

Postman is a chrome extension so there must be a way to do, we just have to find it :)

**Comment by Prudwell** [comment]
Very cool. So far I've just been using base64 encoded text, but my app has future plans for PDFs and images.

I was just playing with the petstore api and uploaded a file, I guess they have a working backend to test on.

[http://petstore.swagger.io/#!/pet/uploadFile](http://petstore.swagger.io/#!/pet/uploadFile)

It does indeed seem to use multipart for the upload

<img width="795" alt="screen shot 2017-03-09 at 12 09 41 pm" src="https://cloud.githubusercontent.com/assets/11758003/23768590/6b5295bc-04c1-11e7-8a61-56655d357ded.png">

**Comment by jamiltz** [comment]
Yes, I was able to upload the attachment multipart request to Sync Gateway following the exact same params as the pet store spec. But then i don't know how to consume it. Sending a GET /{db}/{id}/{attach} requests returns the multipart body but if i've uploaded a PDF to PUT `/{db}/{id}/my_pdf?rev=...` with `Content-Type: application/pdf` I would expect GET `/{db}/{id}/my_pdf` to trigger a download of that PDF file. That's not the case when uploading the file with multipart.

So there's a different issue in each of the two issues discussed here:

- uploading as multipart works but it is not clear how to then retrieve the attachment.
- uploading as string binary works in postman but haven't figured out how to do it in code. I might try sending that request in Java and see if that works.

I've also scheduled some time with @tleyden (from the Sync Gateway team) so we should have things cleared up some time next week.
